File: 2.Add-Two-Numbers/addTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
  function ListNode(val) {
     this.val = val;
     this.next = null;
 }
 
var addTwoNumbers = function(l1, l2) {
    var dummyList = new ListNode(0); // when refering to a linked list, we often refer to its node 
    var sumList = dummyList;
    var carry = 0, x, y,  sum; //x for value in l1, y for that in l2
    
    while(l1 !== null || l2 !== null || carry){
        x = (l1 !== null) ? l1.val : 0;
        y = (l2 !== null) ? l2.val : 0;
        sum = x + y + carry;
        carry = Math.floor(sum / 10);
        sumList.next = new ListNode(sum % 10);
        sumList = sumList.next
        if (l1 !== null) l1 = l1.next;
        if (l2 !== null) l2 = l2.next;
    }
    // console.log(dummyList.next);
    // return dummyList.next;
    
    //put each element of the linked list into an array to meet Leetcode requirement
    var arr =[];
    for(var i = dummyList.next; i !== null; i = i.next){
        arr.unshift(i.next);
    }
    console.log(arr);
    return arr;
};

File: 2.Add-Two-Numbers/addTwoNumbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        dummy = ListNode(0)
        head = dummy
        while l1 or l2:
            if not l1:
                head.next = ListNode((l2.val+carry) % 10)
                carry = (l2.val+carry) / 10
                head = head.next
                l2 = l2.next
            elif not l2:
                head.next = ListNode((l1.val+carry) % 10)
                carry = (l1.val+carry) / 10
                head = head.next
                l1 = l1.next
            else:
                sum = (l1.val+l2.val+carry) % 10
                head.next = ListNode(sum)
                carry = (l1.val+l2.val+carry) / 10
                head = head.next
                l1 = l1.next
                l2 = l2.next
        if carry:
            head.next = ListNode(carry)
        return dummy.next